scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,183 @@
+import logging
+
+from h2.config import H2Configuration
+from h2.connection import H2Connection
+from h2.events import (
+    ConnectionTerminated, DataReceived, ResponseReceived,
+    StreamEnded, StreamReset, WindowUpdated
+)
+from twisted.internet.protocol import connectionDone, Protocol
+
+from scrapy.core.http2.stream import Stream
+from scrapy.http import Request
+
+LOGGER = logging.getLogger(__name__)
+
+
+class H2ClientProtocol(Protocol):
+    # TODO:
+    #  1. Check for user-agent while testing
+    #  2. Add support for cookies
+    #  3. Handle priority updates (Not required)
+    #  4. Handle case when received events have StreamID = 0 (applied to H2Connection)
+    #  1 & 2:
+    #   - Automatically handled by the Request middleware
+    #   - request.headers will have 'Set-Cookie' value
+
+    def __init__(self):
+        config = H2Configuration(client_side=True, header_encoding='utf-8')
+        self.conn = H2Connection(config=config)
+
+        # Address of the server we are connected to
+        # these are updated when connection is successfully made
+        self.destination = None
+
+        # ID of the next request stream
+        # Following the convention made by hyper-h2 each client ID
+        # will be odd.
+        self.next_stream_id = 1
+
+        # Streams are stored in a dictionary keyed off their stream IDs
+        self.streams = {}
+
+        # Boolean to keep track the connection is made
+        # If requests are received before connection is made
+        # we keep all requests in a pool and send them as the connection
+        # is made
+        self.is_connection_made = False
+        self._pending_request_stream_pool = []
```

Comment:
Have you considered using a [deque](https://docs.python.org/3.8/library/collections.html#collections.deque) here?

Reply:
Done :+1: